### ISSUE TITLE
ENH: Add @Cookie, @Header + for more explicit mapping @Body @QueryParam @PathParam

### ISSUE THREAD

**Issue by rbygrave** [dinject/dinject-controller#3]
`@Cookie` to obtain a cookie by name
`@Header` to obtain a header by name

`@Body` to explicitly mark a method parameter as the body content
`@QueryParam` to explicitly mark a method parameter as  a query parameter. Also allow for a default value.
`@PathParam` to explicitly mark a method parameter as a path parameter

**Comment by rbygrave** [comment]
Some thoughts (based on my personal experience with jax-rs annotations):

- I'd personally never use `@PathParam` ... the path parameter always matches the value defined in `@Path` so the annotation just becomes noise.
- There is no `@Body`, we should not need it.
- `@QueryParam` similar to `@PathParam` always should match so in theory we should only strictly need it when we want to define a default value or if the query parameter name is snake case.  Using jax-rs `@QueryParam + @DefaultValue` is ugly ... but we need to make it nice to have default values because is commonly useful.
- `@FormParam` mirrors `@QueryParam`

Hmmm.

**Comment by tipsy** [comment]
Fair enough. I'm not a fan of annotations, but I am a fan of explicit code.
Maybe expanding on the docs and providing examples is sufficient.

**Comment by rbygrave** [comment]
Wow, it's May.  

Adding `@QueryParam`.  This supports cases for:
- Being Explicit
- When the parameter name is snake case or similar (that doesn't match valid java/kotlin parameter name)
- When we want the parameter to have a default value
- For (hopefully shortly) use with the equivilent of jax-rs `@BeanParam`

Then will add `@FormParam`.

**Comment by rbygrave** [comment]
Added `@FormParam`, `@Cookie`, `@Header`.
Also added `@Default` for use with `@FormParam` and `@QueryParam`

```java
 // query param, name is "other" with default value
@QueryParam @Default("42") String other

 // query param with explicit param name "not-path" and type conversion to LocalDate
@QueryParam("not-path") LocalDate notPath,           

// explicit param name and default value
@QueryParam("combo-zero") @Default("43") String combo 

```

**Comment by tipsy** [comment]
Nice. As a heads up, javalin 3 is in release candidate mode.

**Comment by rbygrave** [comment]
Ok, happy with all that.  

Now looking to add a `@Form` which would work in a similar way to JAX-RS `@BeanParam` (but less verbose).

```java
  @Post("other")
  void postFormOther(@Form FooForm fooForm) {

  }
```
Where FooForm is a bean, with properties defaulting to be form parameters.  We can additionally use `@QueryParam`, `@Cookie`, `@Header` on properties.
 
```java
  static class FooForm {

    // public fields    
    public String id;
    public String name;

    // properties with setters
    private String banana;
    public void setBanana(String banana) {
      this.banana = banana;
    }

    // things that are not form parameters ...

    @QueryParam
    public String myQueryParam;

    @Header
    public String myHeader;

    @Cookie
    public String myCookie;
  }
```
The generated code will `new FooForm` and populate it and then pass it to the controller.

Once that is done, we can look to output snippets of swagger json (or yaml) for each method/API (so that we can easily return swagger API docs as a merge / combination of these snippets).

**Comment by rbygrave** [comment]
> Nice. As a heads up, javalin 3 is in release candidate mode.

Yes, great work @tipsy .  I've been watching but unable to keep up with everything unfortunately (I have been busy with ebean.io orm and work etc) ... but I'm keen to push on with this and get the docs up to scratch.

**Comment by rbygrave** [comment]
Ok, adding `@Form` (similar to JAX-RS `@BeanParam`).

## Simple example

```java
public class FooForm {
    public String id;
    public String name;
}
```
```java
  @Post
  void postMyForm(@Form FooForm fooForm) {
    ...
  }
```
So with the above the properties in the FooForm default to being form parameters.  The generated code looks like:

```java
    ApiBuilder.post("/foo/other", ctx -> {
      ctx.status(201);
      FooForm fooForm =  new FooForm();
      fooForm.id = ctx.formParam("id");
      fooForm.name = ctx.formParam("name");

      controller.postFormOther(fooForm, ctx);
    });
```

## More complex example
In this example some of the bean properties have annotations like `@QueryParam`, `@Header`, `@Cookie`, `@Default` ... and we can also use `@FormParam` when we have a form parameter with snake case or similar.

```java
  static class FooForm {

    public String id;

    public String name;

    // interesting form parameter name + default value + type conversion to Integer
    @FormParam("one-but") @Default("42")
    public Integer one;

    // use a setter method + type conversion to UUID
    private UUID banana;

    public void setBanana(UUID banana) {
      this.banana = banana;
    }

    @QueryParam
    public String myQueryParam;

    @Header
    public String myHeader;

    @Cookie
    public String myCookie;

  }
```
And the generated code becomes:
```java
    ApiBuilder.post("/foo/other", ctx -> {
      ctx.status(201);
      FooForm fooForm =  new FooForm();
      fooForm.id = ctx.formParam("id");
      fooForm.name = ctx.formParam("name");
      fooForm.one = toInteger(ctx.formParam("one-but","42"));
      fooForm.setBanana(toUUID(ctx.formParam("banana")));
      fooForm.myQueryParam = ctx.queryParam("myQueryParam");
      fooForm.myHeader = ctx.header("My-Header");
      fooForm.myCookie = ctx.cookie("myCookie");

      controller.postFormOther(fooForm, ctx);
    });
```
